cloud/services/scalesets/vmss.go

```diff
@@ -222,7 +220,41 @@ func (s *Service) Reconcile(ctx context.Context) error {
 			return errors.Wrapf(err, "cannot create VMSS")
 		}
 		s.Scope.V(2).Info("successfully created VMSS", "scale set", vmssSpec.Name)
+		s.Scope.SaveK8sVersion()
+	}
+
+	// get the VMSS to update status
+	existingVMSS, err := s.getExisting(ctx, vmssSpec.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get VMSS %s after create or update", vmssSpec.Name)
+	}
+
+	// check to see if we are running the most K8s version specified in the MachinePool spec
+	// if not, then update the instances that are not running that model
+	if s.Scope.NeedsK8sVersionUpdate() {
```

Comment:
A few questions:

1) wouldn't the above update already take care of updating the model?
2) do we want to limit VMSS rolling updates to k8s versions? It might be interesting for other use cases as well, just like machine deployments and KCP allow rolling updates for other things (eg. changing the AzureMachine VM size).
3) we need to define what we consider mutable because CAPI assumes immutable machines but this would break that principle by updating the instances instead of replacing them https://cluster-api.sigs.k8s.io/user/concepts.html?highlight=immutable#machine-immutability-in-place-upgrade-vs-replace

Reply:
1. Yes the above update will have updated the model in Azure to the new version. The status of the AzureMachinePool has not been updated to reflect that yet. The status is still `version - 1` and the MachinePool spec is `version`. At this point, we are checking if we need to cause the instances to update to the new model version.
2. Agreed. I constrained it for the time being since I think we need to work on this behavior. Also, I don't believe this is a rolling upgrade. I think it will upgrade all of the instances at once. I'm not super hyped about that and would be willing to dig into this more to implement rolling upgrade. Thoughts?
3. These machines are not mutable. It's just they may have the same name with a new instance. I'm pretty sure the node has a different ID, just the provider ID might not change. I think this is still immutable. Perhaps, we need to have a unique provider ID?